File: runtime/python/src/mica/cli/graph.py
from __future__ import annotations

from mica.cli.validate import LoadedApp


def graph_ascii(loaded: LoadedApp) -> str:
    names = [proc.spec_name for proc in loaded.processes]
    languages = {
        proc.spec_name: proc.component.language.upper() if proc.component.language != "cpp" else "C++"
        for proc in loaded.processes
    }
    if loaded.processes and loaded.processes[0].component.language == "python":
        languages[loaded.processes[0].spec_name] = "Python"
    for proc in loaded.processes:
        if proc.component.language == "python":
            languages[proc.spec_name] = "Python"
        elif proc.component.language == "go":
            languages[proc.spec_name] = "Go"
        else:
            languages[proc.spec_name] = "C++"

    width = max(len(name) for name in names + [" "]) + 2
    width = max(width, 18)

    def box(name: str) -> list[str]:
        label = languages[name]
        inner = width
        border = "+" + "-" * inner + "+"
        return [
            border,
            "|" + name.ljust(inner) + "|",
            "|" + label.ljust(inner) + "|",
            border,
        ]

    edges: list[tuple[str, str, str]] = []
    publishers: dict[str, list[str]] = {}
    subscribers: dict[str, list[str]] = {}
    callers: dict[str, list[str]] = {}
    providers: dict[str, list[str]] = {}
    for proc in loaded.processes:
        for item in proc.component.publishes:
            publishers.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.subscribes:
            subscribers.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.calls:
            callers.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.provides:
            providers.setdefault(item, []).append(proc.spec_name)
    for contract, pubs in publishers.items():
        for src in pubs:
            for dst in subscribers.get(contract, []):
                edges.append((src, dst, contract))
    for contract, srcs in callers.items():
        for src in srcs:
            for dst in providers.get(contract, []):
                edges.append((src, dst, f"RPC {contract}"))

    lines: list[str] = []
    boxed: set[str] = set()
    if not names:
        return ""

    def ensure_box(name: str) -> None:
        if name in boxed:
            return
        for line in box(name):
            lines.append(line)
        boxed.add(name)

    for name in names:
        ensure_box(name)
        outgoing = [edge for edge in edges if edge[0] == name]
        for _, dst, label in outgoing:
            lines.append(" " * (width // 2) + "|")
            lines.append(" " * (width // 2) + f"| {label}")
            lines.append(" " * (width // 2) + "v")
            if dst in boxed:
                lines.append(" " * ((width - len(dst)) // 2) + dst)
            else:
                ensure_box(dst)
        if outgoing:
            lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def graph_dot(loaded: LoadedApp) -> str:
    lines = [f'digraph "{loaded.app.name}" {{', "  rankdir=TB;"]
    for proc in loaded.processes:
        lang = proc.component.language
        lines.append(
            f'  "{proc.spec_name}" [label="{proc.spec_name}\\n{lang}"];'
        )
    seen: set[tuple[str, str, str]] = set()
    pubs: dict[str, list[str]] = {}
    subs: dict[str, list[str]] = {}
    calls: dict[str, list[str]] = {}
    prov: dict[str, list[str]] = {}
    for proc in loaded.processes:
        for item in proc.component.publishes:
            pubs.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.subscribes:
            subs.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.calls:
            calls.setdefault(item, []).append(proc.spec_name)
        for item in proc.component.provides:
            prov.setdefault(item, []).append(proc.spec_name)
    for contract, srcs in pubs.items():
        for src in srcs:
            for dst in subs.get(contract, []):
                key = (src, dst, contract)
                if key in seen:
                    continue
                seen.add(key)
                lines.append(f'  "{src}" -> "{dst}" [label="{contract}"];')
    for contract, srcs in calls.items():
        for src in srcs:
            for dst in prov.get(contract, []):
                key = (src, dst, contract)
                if key in seen:
                    continue
                seen.add(key)
                lines.append(f'  "{src}" -> "{dst}" [label="RPC {contract}"];')
    lines.append("}")
    return "\n".join(lines) + "\n"

File: cli/internal/graph/graph.go
package graph

import (
	"fmt"
	"strings"

	"mica/cli/internal/manifest"
	"mica/cli/internal/workspace"
)

func ASCII(ws *workspace.Workspace) string {
	type node struct {
		process   string
		component *manifest.Component
	}
	var nodes []node
	for _, process := range ws.App.Processes {
		component, ok := ws.ProcessComponent(process)
		if !ok {
			continue
		}
		nodes = append(nodes, node{process: process.Name, component: component})
	}
	if len(nodes) == 0 {
		return ""
	}
	width := 18
	for _, item := range nodes {
		if len(item.process)+2 > width {
			width = len(item.process) + 2
		}
	}
	languages := map[string]string{}
	for _, item := range nodes {
		languages[item.process] = languageLabel(item.component.Language)
	}
	type edge struct {
		src, dst, label string
	}
	var edges []edge
	for _, item := range nodes {
		for _, contract := range item.component.Publishes {
			for _, other := range nodes {
				if has(other.component.Subscribes, contract) {
					edges = append(edges, edge{item.process, other.process, contract})
				}
			}
		}
		for _, contract := range item.component.Calls {
			for _, other := range nodes {
				if has(other.component.Provides, contract) {
					edges = append(edges, edge{item.process, other.process, "RPC " + contract})
				}
			}
		}
	}
	box := func(name string) []string {
		border := "+" + strings.Repeat("-", width) + "+"
		return []string{
			border,
			"|" + pad(name, width) + "|",
			"|" + pad(languages[name], width) + "|",
			border,
		}
	}
	var lines []string
	boxed := map[string]bool{}
	ensureBox := func(name string) {
		if boxed[name] {
			return
		}
		lines = append(lines, box(name)...)
		boxed[name] = true
	}
	for _, item := range nodes {
		ensureBox(item.process)
		outgoing := []edge{}
		for _, candidate := range edges {
			if candidate.src == item.process {
				outgoing = append(outgoing, candidate)
			}
		}
		for _, candidate := range outgoing {
			indent := strings.Repeat(" ", width/2)
			lines = append(lines, indent+"|")
			lines = append(lines, indent+"| "+candidate.label)
			lines = append(lines, indent+"v")
			if boxed[candidate.dst] {
				lines = append(lines, strings.Repeat(" ", max(0, (width-len(candidate.dst))/2))+candidate.dst)
			} else {
				ensureBox(candidate.dst)
			}
		}
		if len(outgoing) > 0 {
			lines = append(lines, "")
		}
	}
	return strings.TrimRight(strings.Join(lines, "\n"), "\n") + "\n"
}

func DOT(ws *workspace.Workspace) string {
	var out strings.Builder
	fmt.Fprintf(&out, "digraph %q {\n", ws.App.Name)
	out.WriteString("  rankdir=TB;\n")
	for _, process := range ws.App.Processes {
		component, ok := ws.ProcessComponent(process)
		if !ok {
			continue
		}
		fmt.Fprintf(&out, "  %q [label=%q];\n", process.Name, process.Name+"\\n"+component.Language)
	}
	seen := map[string]bool{}
	add := func(src, dst, label string) {
		key := src + "\x00" + dst + "\x00" + label
		if seen[key] {
			return
		}
		seen[key] = true
		fmt.Fprintf(&out, "  %q -> %q [label=%q];\n", src, dst, label)
	}
	for _, process := range ws.App.Processes {
		component, ok := ws.ProcessComponent(process)
		if !ok {
			continue
		}
		for _, contract := range component.Publishes {
			for _, other := range ws.App.Processes {
				otherComponent, ok := ws.ProcessComponent(other)
				if ok && has(otherComponent.Subscribes, contract) {
					add(process.Name, other.Name, contract)
				}
			}
		}
		for _, contract := range component.Calls {
			for _, other := range ws.App.Processes {
				otherComponent, ok := ws.ProcessComponent(other)
				if ok && has(otherComponent.Provides, contract) {
					add(process.Name, other.Name, "RPC "+contract)
				}
			}
		}
	}
	out.WriteString("}\n")
	return out.String()
}

func has(items []string, value string) bool {
	for _, item := range items {
		if item == value {
			return true
		}
	}
	return false
}

func languageLabel(language string) string {
	switch language {
	case "python":
		return "Python"
	case "go":
		return "Go"
	case "cpp":
		return "C++"
	}
	return language
}

func pad(value string, width int) string {
	if len(value) >= width {
		return value
	}
	return value + strings.Repeat(" ", width-len(value))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
